pulsar-client/src/main/java/org/apache/pulsar/client/impl/AutoClusterFailover.java

```diff
@@ -0,0 +1,234 @@
+                                    + "switch back to primary service: {}", currentPulsarServiceUrl,
+                                    System.currentTimeMillis() - secondaryFailedTimestamp, primary);
+                            try {
+                                pulsarClient.updateServiceUrl(primary);
+                                currentPulsarServiceUrl = primary;
+                                return;
+                            } catch (PulsarClientException e) {
+                                log.error("Current Pulsar service is secondary: {}, it has been down for {} ms. "
+                                        + "Failed to switch to secondary service URL ",
+                                        currentPulsarServiceUrl,
+                                        System.currentTimeMillis() - secondaryFailedTimestamp, e);
+                            }
+                        }
+
+                        return;
+                    }
+
+                    secondaryFailedTimestamp = -1;
+
+                    if (!probeAvailable(primary, timeout)) {
+                        primaryRecoverTimestamp = -1;
+                        return;
+                    }
+                    if (primaryRecoverTimestamp == -1) {
+                        primaryRecoverTimestamp = System.currentTimeMillis();
+                    } else if (System.currentTimeMillis() - primaryRecoverTimestamp >= switchBackDelayMs) {
+                        log.info("Current Pulsar service is secondary: {}, "
+                                        + "the primary service: {} has been recover for {} ms, "
+                                        + "switch back to the primary service",
+                                currentPulsarServiceUrl, primary, System.currentTimeMillis() - primaryRecoverTimestamp);
+                        try {
+                            pulsarClient.updateServiceUrl(primary);
+                            currentPulsarServiceUrl = primary;
+                        } catch (PulsarClientException e) {
+                            log.error("Current Pulsar service is secondary: {}, "
+                                    + "failed to switch back to primary service URL ", currentPulsarServiceUrl, e);
+                        }
+                    }
+                }
+            }
+        }, 30_000, 30_000);
+
+    }
+
+    @Override
+    public String getServiceUrl() {
+        return this.currentPulsarServiceUrl;
+    }
+
+    @Override
+    public void close() {
+        this.timer.cancel();
+    }
+
+    private boolean probeAvailable(String url, int timeout) {
+        try {
+            String hostAndPort = parseHostAndPort(url);
+            if (Strings.isNullOrEmpty(hostAndPort)) {
+                return false;
+            }
+
+            Socket socket = new Socket();
+            socket.connect(new InetSocketAddress(parseHost(hostAndPort), parsePort(hostAndPort)), timeout);
```

Comment:
It might be better to make a request to the server to ensure it is actually in a good state.

Reply:
The current implementation is probe whether the service port is open or not. If we want to probe the broker state, it's better to provide a health check command for broker service on broker side. I can use another PR to provide this feature on broker side. @merlimat Do you have any other ideas?